### PR TITLE
KAFKA-6074 Use ZookeeperClient in ReplicaManager and Partition

### PR DESCRIPTION
**Pull request by tedyu**

### CODE REVIEW

**Comment by tedyu** [comment]
In zkUtils.conditionalUpdatePersistentPath():
```
      val stat = zkClientWrap(_.writeDataReturnStat(path, data, expectVersion))
```
Currently there is no writeData in KafkaControllerZkUtils.
Planning to add it.

See if what I have so far is on right track.

**Comment by tedyu** [comment]
More compilation errors to tackle:

https://pastebin.com/rK0FwQBJ

**Comment by omkreddy** [comment]
@tedyu  I also ported the conditionalUpdatePersistentPath method in my #4126 PR. Pl check if we can reuse that code .

File: core/src/main/scala/kafka/controller/ZookeeperClient.scala [hunk 1]
```diff
@@ -361,6 +365,7 @@ sealed trait AsyncResponse {
 }
 case class CreateResponse(resultCode: Code, path: String, ctx: Option[Any], name: String) extends AsyncResponse
 case class DeleteResponse(resultCode: Code, path: String, ctx: Option[Any]) extends AsyncResponse
+case class WriteDataReturnStatResponse(resultCode: Code, path: String, ctx: Option[Any], stat: Stat) extends AsyncResponse
```
Comment (omkreddy) [inline core/src/main/scala/kafka/controller/ZookeeperClient.scala:368]: we can use SetDataResponse for writing right? Pls check #4126

**Comment by onurkaraman** [comment]
Just a heads up: I consciously avoided copying existing util methods from ZkUtils over to KafkaControllerZkUtils.

**Comment by tedyu** [comment]
I will wait till #4126 gets merged (for code reuse)

File: core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala [hunk 2]
```diff
@@ -347,6 +354,18 @@ class KafkaControllerZkUtils(zookeeperClient: ZookeeperClient, isSecure: Boolean
     retryRequestUntilConnected(setDataRequest)
   }
 
+  def createSequentialPersistentPath(path: String, data: String = ""): String = {
+    val createRequest = CreateRequest(path, data.getBytes("UTF-8"), acls(path), CreateMode.PERSISTENT_SEQUENTIAL)
+    val createResponse = retryRequestUntilConnected(createRequest)
+    createResponse.path
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala:360]: We want to throw an exception if the rc in createResponse is not Code.OK.

File: core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala [hunk 3]
```diff
@@ -98,6 +100,11 @@ class KafkaControllerZkUtils(zookeeperClient: ZookeeperClient, isSecure: Boolean
     retryRequestUntilConnected(createRequest)
   }
 
+  def leaderAndIsrZkData(leaderAndIsr: LeaderAndIsr, controllerEpoch: Int): String = {
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala:103]: This probably should be in ZkData.scala.

**Review by junrao — COMMENTED** [review]
@tedyu : Thanks for the patch. A couple of comments below.

**Comment by tedyu** [comment]
Thanks for the review, Jun.

Planning to rebase once #4126 goes in.

**Comment by ijuma** [comment]
This was superseded by https://github.com/apache/kafka/pull/4254

**Pull request closed without merging**